server/src/main/java/org/opensearch/plugins/Plugin.java

```diff
@@ -119,25 +121,37 @@ public Collection<Class<? extends LifecycleComponent>> getGuiceServiceClasses()
         return Collections.emptyList();
     }
 
+    /**
+     * Setter for PluginSubject.
+     *
+     * @param pluginSubject A subject for executing transport actions as the plugin
+     */
+    final void setPluginSubject(PluginSubject pluginSubject) {
```

Comment:
I think we should move it away from the `Plugin` and bundle it with `IdentityPlugin` instead by introducing something like `IdentityAwarePlugin`. It will be kicking in **only** when `IdentityPlugin` is available.

Reply:
I think `IdentityAwarePlugin` would be 1-to-1 with `SystemIndexPlugin`. Would it make sense to add this setter there?

Inside the [`IdentityService`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/identity/IdentityService.java#L35) if no identity plugins are installed then it defaults to `NoopIdentityPlugin`. Does it make sense to use the `PluginSubject` with the `NoopIdentityPlugin`. I didn't intend to have this code marked experimental. What would be the criteria for marking it out of experimental?

Reply:
> I think `IdentityAwarePlugin` would be 1-to-1 with `SystemIndexPlugin`. Would it make sense to add this setter there?

The `Subject` is part of the  `IdentityPlugin` , to me the  `SystemIndexPlugin` and   `IdentityPlugin`  are very unrelated to be fair.

> Inside the [IdentityService](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/identity/IdentityService.java#L35) if no identity plugins are installed then it defaults to NoopIdentityPlugin. Does it make sense to use the PluginSubject with the NoopIdentityPlugin

The `PluginSubject` is static plugin identity, right? So we could have `NoopPluginSubject` fe to emphasize that this is noop impl.

Reply:
Got it, I just pushed a commit to address all comments and have drastically simplified this PR. This will now delegate all implementation to the IdentityPlugin and use a Noop implementation if no IdentityPlugin is installed.

Thank you for all the regular feedback @reta ! I greatly appreciate it.

As a follow-up to this PR, I want to take IdentityPlugin out of experimental in 2.18. Do you think that would be feasible?

Reply:
> As a follow-up to this PR, I want to take IdentityPlugin out of experimental in 2.18. Do you think that would be feasible?

I sadly not sure, we probably need to have at least one full fledged implementation (better more than one) to make sure we cover at least necessary functionality, but we could discuss for sure.